go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,484 @@
+	}
+	if !strings.Contains(output, successMsg) {
+		return mysql.Position{}, vterrors.Errorf(vtrpc.Code_UNKNOWN, "copy-back step failed: %v", output)
+	}
+
+	// now find the slave position and return that
+	logger.Infof("Returning replication position %v", bm.Position)
+	// TODO clean up extracted files from TmpDir
+	return bm.Position, nil
+}
+
+// restoreFile restores an individual file.
+func (be *XtrabackupEngine) restoreFile(
+	ctx context.Context,
+	cnf *Mycnf,
+	logger logutil.Logger,
+	bh backupstorage.BackupHandle,
+	compress bool,
+	name string) (err error) {
+
+	streamMode := *xtrabackupStreamMode
+	// Open the source file for reading.
+	var source io.ReadCloser
+	source, err = bh.ReadFile(ctx, name)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+
+	reader := io.MultiReader(source)
+
+	// Create the uncompresser if needed.
+	if compress {
+		gz, err := pgzip.NewReader(reader)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if cerr := gz.Close(); cerr != nil {
+				if err != nil {
+					// We already have an error, just log this one.
+					logger.Errorf("failed to close gunziper %v: %v", name, cerr)
+				} else {
+					err = cerr
+				}
+			}
+		}()
+		reader = gz
+	}
+
+	switch streamMode {
+	case streamModeTar:
+		// now extract the files by running tar
+		// error if we can't find tar
+		flagsToExec := []string{"-C", cnf.TmpDir, "-xi"}
+		tarCmd := exec.Command("tar", flagsToExec...)
+		logger.Infof("Executing tar cmd with flags %v", flagsToExec)
+		tarCmd.Stdin = reader
+		tarOut, _ := tarCmd.StdoutPipe()
+		tarErr, _ := tarCmd.StderrPipe()
+		tarCmd.Start()
+		output, _ := ioutil.ReadAll(tarOut)
+		errOutput, _ := ioutil.ReadAll(tarErr)
+		err := tarCmd.Wait()
+
+		if string(output) != "" {
+			logger.Infof("output from tar: %v ", string(output))
+		}
+		if string(errOutput) != "" {
+			logger.Infof("error from tar: %v ", string(errOutput))
+		}
+		if err != nil {
+			return vterrors.Wrap(err, "error from tar")
+		}
+
+	case xbstream:
+		// now extract the files by running xbstream
+		xbstreamProgram := xbstream
+		flagsToExec := []string{}
+		if *xbstreamRestoreFlags != "" {
+			flagsToExec = append(flagsToExec, strings.Fields(*xbstreamRestoreFlags)...)
+		}
+		flagsToExec = append(flagsToExec, "-C", cnf.TmpDir, "-x")
+		xbstreamCmd := exec.Command(xbstreamProgram, flagsToExec...)
+		logger.Infof("Executing xbstream cmd: %v %v", xbstreamProgram, flagsToExec)
+		xbstreamCmd.Stdin = reader
+		xbstreamOut, _ := xbstreamCmd.StdoutPipe()
+		xbstreamErr, _ := xbstreamCmd.StderrPipe()
+		xbstreamCmd.Start()
+		output, _ := ioutil.ReadAll(xbstreamOut)
+		errOutput, _ := ioutil.ReadAll(xbstreamErr)
+		err := xbstreamCmd.Wait()
+
+		if string(output) != "" {
+			logger.Infof("Output from xbstream: %v ", string(output))
+		}
+		if string(errOutput) != "" {
+			logger.Infof("error from xbstream: %v", string(errOutput))
+		}
+		if err != nil {
+			return vterrors.Wrap(err, "error from xbstream")
+		}
+	default:
+	}
```

Comment:
add some err to panic or return unsupported?

Reply:
done